### ISSUE TITLE
Performance issues

### ISSUE THREAD

**Issue by tarcieri** [rust-bitcoin/rust-secp256k1#40]
Hello, I'm the author of a multi-provider Rust digital signature library called Signatory:

https://github.com/tendermint/signatory

I've added an ECDSA provider for the `secp256k1` crate here:

https://github.com/tendermint/signatory/blob/master/src/providers/secp256k1/ecdsa.rs

and a benchmark here:

https://github.com/tendermint/signatory/blob/master/benches/secp256k1.rs

Benchmarking it side-by side with several other elliptic curve digital signature libraries, the `secp256k1` crate seems unexpectedly slow, like compiler optimizations aren't being applied or something.

Running the benchmark suite on a Intel Xeon E3-1225 v5 @ 3.30GHz with `-Ctarget_cpu=native` I get roughly:

* ed25519-dalek (AVX2 backend): 16us sign, 34us verify
* ring (Ed25519): 21us sign, 62us verify
* ring (ECDSA P-256): 18us sign, 56us verify
* secp256k1: 111us sign, 123us verify

You can run the same suite by cloning the Signatory repo and running `RUSTFLAGS=-Ctarget_cpu=native cargo bench --all-features`

Among other things it looks like you're using the `gcc` crate, which has been superseded by the `cc` crate. But are there optimization flags which should be supplied which aren't?

**Comment by tarcieri** [comment]
Results from `bench_sign` and `bench_verify` w\ `CFLAGS=-march=native`:

```
ecdsa_sign: min 41.0us / avg 41.1us / max 41.5us
ecdsa_verify: min 66.1us / avg 66.2us / max 66.2us
ecdsa_verify_openssl: min 382us / avg 382us / max 383us
```

@apoelstra FWIW I believe *ring* is using [fiat-crypto](https://github.com/mit-plv/fiat-crypto). Since their main use case is TLS (I believe, anyway), it's possible they've spent more time optimizing signing than verification.

**Comment by apoelstra** [comment]
Ok, great, those sign and verify times look correct to me (I think verification will drop to about 45 us if you use `--enable-endomorphism` on your configure line).

I believe the bulk of the slowdown is that rust-secp256k1 is using the 32-bit version of libsecp256k1, which is a bug in the bindings, not the underlying library (or a quirk of your setup).

We are gearing up for a 1.0 release of these bindings. I think this bug is blocking that.

**Comment by tarcieri** [comment]
Results with `--enable-endomorphism`:

```
ecdsa_sign: min 40.8us / avg 40.9us / max 41.0us
ecdsa_verify: min 50.3us / avg 50.4us / max 50.4us
ecdsa_verify_openssl: min 381us / avg 382us / max 382us
```

**Comment by apoelstra** [comment]
Fixed by #42
